src/utils/internal/validate.ts

```diff
@@ -0,0 +1,33 @@
+import { createError } from "../../error";
+
+// TODO: Consider using similar method of typeschema for external library compatibility
+// https://github.com/decs/typeschema/blob/v0.1.3/src/assert.ts
+
+export type ValidateResult<T> = T | false | void;
+
+export type ValidateFunction<T> = (
+  data: unknown
+) => ValidateResult<T> | Promise<ValidateResult<T>>;
+
+export async function validateData<T>(
+  data: unknown,
+  fn: ValidateFunction<T>
+): Promise<T> {
+  try {
+    const res = await fn(data);
+    if (res === false) {
+      throw createValidationError();
+    }
+    return res ?? (data as T);
+  } catch (error) {
+    throw createValidationError(error);
+  }
+}
+
+function createValidationError(validateError?: any) {
+  throw createError({
+    status: 400,
```

Comment:
We can consider 422 here https://www.rfc-editor.org/rfc/rfc9110.html#name-422-unprocessable-content

Reply:
I assume 400-bad request is more appropriate for default assumption. 422 means data is semantically correct while it might not be as a result of a failed validation (string passed instead of an object, wrong nested object, etc)